### PR TITLE
Yoshi Serverless POC

### PR DESCRIPTION
**Pull request by yanivefraim**
This PR is a POC for Yoshi Serverless:
- added support for serverless as an entry point
- dynamically create a serverless folder
- transpile server files to serverless folder
- everything is under the `EXPERIMENTAL_YOSHI_SERVERLESS` flag

### CODE REVIEW

File: packages/yoshi-serverless/src/routes/api.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
+import path from 'path';
+import globby from 'globby';
+import { isLeft } from 'fp-ts/lib/Either';
+import { BUILD_DIR } from 'yoshi-config/build/paths';
+import { requestPayloadCodec, DSL } from '../types';
+import { relativeFilePath, connectToYoshiServerHMR } from '../utils';
+import { route } from '..';
+
+const buildDir = path.resolve(__dirname, '..', '..', BUILD_DIR);
```
Comment (yanivefraim) [inline packages/yoshi-serverless/src/routes/api.ts:9]: do we really need these?

File: packages/yoshi-flow-app/src/scripts/build.ts [hunk 2]
```diff
@@ -63,6 +64,7 @@ const build: cliCommand = async function(argv, config) {
   await Promise.all([
     fs.emptyDir(join(BUILD_DIR)),
     fs.emptyDir(join(TARGET_DIR)),
+    fs.emptyDir(join('serverless')),
```
Comment (ronami) [inline packages/yoshi-flow-app/src/scripts/build.ts:67]: `SERVERLESS_DIR`

File: packages/yoshi-common/src/copy-serverless.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  const serverlessTemplateFolder = path.resolve(
+    __dirname,
+    '..',
+    'src',
+    `./templates/serverless`,
+  );
+
+  const serverlessDestinationPath = path.resolve(
+    cwd,
+    `serverless/${projectName}`,
```
Comment (ronami) [inline packages/yoshi-common/src/copy-serverless.ts:23]: `SERVERLESS_DIR`

File: packages/yoshi-common/src/copy-serverless.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  if (!config.yoshiServer || !process.env.EXPERIMENTAL_YOSHI_SERVERLESS) {
+    return;
+  }
+
+  const serverlessTemplateFolder = path.resolve(
+    __dirname,
+    '..',
+    'src',
+    `./templates/serverless`,
+  );
```
Comment (ronami) [inline packages/yoshi-common/src/copy-serverless.ts:19]: super tiny but can be consistent to either separate with `,` or `/`, also `SERVERLESS_DIR`

File: packages/yoshi-common/src/templates/serverless/serverless.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+const { Server } = require('yoshi-serverless');
+
+module.exports = functionsBuilder => {
+  return functionsBuilder
+    .addWebFunction('POST', '*', async (ctx, req) => {
+      const server = await Server.create(ctx);
+      const a = await server.handle(req);
```
Comment (ronami) [inline packages/yoshi-common/src/templates/serverless/serverless.js:7]: Probably `handler` explains it better

File: packages/yoshi-common/src/webpack-utils.ts [hunk 6]
```diff
@@ -124,7 +124,10 @@ function createServerEntries(context: string, cwd: string = process.cwd()) {
   }, {});
 
   // Add custom entries for `yoshi-server`
-  entries['routes/_api_'] = 'yoshi-server/build/routes/api';
+  entries['routes/_api_'] = `yoshi-server${
+    process.env.EXPERIMENTAL_YOSHI_SERVERLESS ? 'less' : ''
```
Comment (ronami) [inline packages/yoshi-common/src/webpack-utils.ts:128]: A bit confusing IMO, I think `if () else {}` might be a bit more clear

File: packages/yoshi-common/src/webpack.config.ts [hunk 7]
```diff
@@ -495,7 +496,11 @@ export function createBaseWebpackConfig({
 
       ...(target === 'node'
         ? {
-            path: join(BUILD_DIR),
+            path: join(
+              process.env.EXPERIMENTAL_YOSHI_SERVERLESS
+                ? `${SERVERLESS_DIR}/${getProjectArtifactId()}/dist`
```
Comment (ronami) [inline packages/yoshi-common/src/webpack.config.ts:501]: I think we can move this to `paths.js` so it's not duplicated

File: packages/yoshi-flow-app/src/scripts/build.ts [hunk 8]
```diff
@@ -71,6 +73,7 @@ const build: cliCommand = async function(argv, config) {
     const petriSpecs = await import('yoshi-common/build/sync-petri-specs');
     const wixMavenStatics = await import('yoshi-common/build/maven-statics');
     const copyDocker = await import('yoshi-common/build/copy-docker');
+    const copyServerless = await import('yoshi-common/build/copy-serverless');
```
Comment (ronami) [inline packages/yoshi-flow-app/src/scripts/build.ts:76]: Might be easier with plugin so it happens automatically for `start` and doesn't require flows to add this bit of code.

File: packages/yoshi-server-tools/src/utils.ts [hunk 9]
```diff
@@ -28,7 +28,11 @@ export function transform(source: string, fullFileName: string) {
   const srcDir = path.resolve(SRC_DIR);
   const fileName = path.relative(srcDir, fullFileName).replace(/\.[^/.]+$/, '');
 
-  const headers = [`import { dsl } from 'yoshi-server/build/wrap';`];
+  const headers = [
+    `import { dsl } from 'yoshi-server${
+      process.env.EXPERIMENTAL_YOSHI_SERVERLESS ? 'less' : ''
```
Comment (ronami) [inline packages/yoshi-server-tools/src/utils.ts:33]: might be a bit more clear if we construct it in a var outside

File: packages/yoshi-serverless/package.json [hunk 10]
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "yoshi-serverless",
+  "version": "4.83.0",
+  "main": "build/index.js",
+  "sideEffects": false,
+  "keywords": [
+    "yoshi",
+    "server"
+  ],
+  "author": "Ronen Amiel",
```
Comment (ronami) [inline packages/yoshi-serverless/package.json:10]: Should be `Yaniv Efraim` :)

File: packages/yoshi-serverless/src/routes/api.ts [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      functions = createFunctions();
+    } catch (error) {
+      socket.send(JSON.stringify({ success: false }));
+    }
+    socket.send(JSON.stringify({ success: true }));
+  };
+}
+
+export default route(async function() {
+  const body = this.req.body; // || (await parseBodyAsJson(this.req));
```
Comment (ronami) [inline packages/yoshi-serverless/src/routes/api.ts:47]: we should check if this is needed

File: packages/yoshi-serverless/src/routes/launchEditor.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import launchEditor from 'react-dev-utils/launchEditor';
+import { route } from '..';
+
+export default route(async function() {
+  launchEditor(
+    '/' + this.req.query.file,
+    parseInt(this.req.query.line as string, 10),
+    parseInt(this.req.query.col as string, 10),
+  );
+});
```
Comment (ronami) [inline packages/yoshi-serverless/src/routes/launchEditor.ts:10]: might want to remove until we need it

File: packages/yoshi-serverless/src/server.ts [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    } catch (error) {
+      // if (
+      //   process.env.NODE_ENV === 'production' ||
+      //   process.env.IS_INTEGRATION_TEST_PROD === 'true'
+      // ) {
+      //   return next(new InternalServerError('internal server error', error));
+      // }
+      // const youch = new Youch(error, req);
+      // const html: string = await youch.toHTML();
+      // return send(res, 500, html);
```
Comment (ronami) [inline packages/yoshi-serverless/src/server.ts:92]: we can return a string of `500` or `400` like we do in the route so its easier for us to debug

File: packages/yoshi-serverless/src/server.ts [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    // if (process.env.NODE_ENV === 'production') {
+    //   // If Yoshi Server did not find anything, pass the request on
+    //   return next();
+    // }
+
+    // return send(res, 404, 'not found');
```
Comment (ronami) [inline packages/yoshi-serverless/src/server.ts:100]: same as top

File: packages/yoshi-serverless/src/types.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+// General stuff
+export type UnpackPromise<T> = T extends Promise<infer U> ? U : T;
+export type OptionalPromise<T> = T | Promise<T>;
+export type FunctionArgs = Array<any>;
+export type FunctionResult = OptionalPromise<any>;
+
+// Server function types
+export type FunctionContext = {
+  req: WebRequest; // & WithAspects;
+  context: any;
```
Comment (ronami) [inline packages/yoshi-serverless/src/types.ts:23]: I'm not sure we have a context, if we do I think it's a different context, does it have types?

File: packages/yoshi-serverless/src/types.ts [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+  initData: any;
+  config: any;
```
Comment (ronami) [inline packages/yoshi-serverless/src/types.ts:25]: Not sure if we have this

File: packages/yoshi-flow-app/src/scripts/build.ts (hunk 2 above)
Comment (yanivefraim) [reply packages/yoshi-flow-app/src/scripts/build.ts:67]: done

File: packages/yoshi-common/src/copy-serverless.ts (hunk 3 above)
Comment (yanivefraim) [reply packages/yoshi-common/src/copy-serverless.ts:23]: done

File: packages/yoshi-flow-app/src/scripts/build.ts (hunk 8 above)
Comment (yanivefraim) [reply packages/yoshi-flow-app/src/scripts/build.ts:76]: let's do it in a different PR, together with other copying tasks

Comment (yanivefraim) [reply packages/yoshi-flow-app/src/scripts/build.ts:80]: Let's do it in a different PR, with other copy tasks

**Pull request merged**